rocAL/rocAL/source/pipeline/master_graph.cpp

```diff
@@ -535,428 +428,29 @@ MasterGraph::mem_type()
-        return MasterGraph::Status::INVALID_ARGUMENTS;
-
-    _convert_time.start();
-
-#if ENABLE_OPENCL
-    if(processing_on_device_ocl())
-    {
-        size_t dest_buf_offset = 0;
-        //NOTE: the CL_TRUE flag is only used on the last buffer read call,
-        // to avoid unnecessary sequence of synchronizations
-
-        // get_read_buffers() calls block_if_empty() internally and blocks if buffers are empty until a new batch is processed
-        auto output_buffers =_ring_buffer.get_read_buffers();
-        auto out_image_idx = output_buffers.size();
-        for( auto&& output_handle: output_buffers)
-        {
-            bool sync_flag = (--out_image_idx == 0) ? CL_TRUE : CL_FALSE;
-            cl_int status;
-            if((status = clEnqueueReadBuffer(_device.resources()->cmd_queue,
-                                             (cl_mem) output_handle,
-                                             sync_flag?(CL_TRUE):CL_FALSE,
-                                             0,
-                                             size,
-                                             out_ptr+dest_buf_offset,
-                                             0 , nullptr, nullptr)) != CL_SUCCESS)
-                THROW("clEnqueueReadBuffer failed: " + TOSTR(status))
-            dest_buf_offset += size;
-        }
-    }
-    else {
-#elif ENABLE_HIP
-    if(processing_on_device_hip())
-    {
-        //NOTE: the CL_TRUE flag is only used on the last buffer read call,
-        // to avoid unnecessary sequence of synchronizations
-
-        // get_read_buffers() calls block_if_empty() internally and blocks if buffers are empty until a new batch is processed
-        size_t dest_buf_offset = 0;
-        auto output_buffers =_ring_buffer.get_read_buffers();
-        for( auto&& output_handle: output_buffers)
-        {
-            hipError_t err = hipMemcpyDtoHAsync((void *)(out_ptr+dest_buf_offset), output_handle, size, _device.resources()->hip_stream);
-            if (err) {
-                THROW("hipMemcpyDtoHAsync failed: " + TOSTR(err))
-            }
-            dest_buf_offset += size;
-        }
-        // sync to finish copy
-        if (hipStreamSynchronize(_device.resources()->hip_stream) != hipSuccess)
-            THROW("hipStreamSynchronize failed for hipMemcpy ")
-
-    }
-    else {
-#endif
-        // get_host_master_read_buffer is blocking if _ring_buffer is empty, and blocks this thread till internal processing thread process a new batch and store in the _ring_buffer
-        memcpy(out_ptr, _ring_buffer.get_host_master_read_buffer(), size * _output_images.size());
-#if ENABLE_OPENCL || ENABLE_HIP
-    }
-#endif    
-    _convert_time.end();
-    return Status::OK;
+    std::vector<void*> output_ptr = _ring_buffer.get_read_buffers();
+    for(unsigned i = 0; i < _internal_tensor_list.size(); i++)
+        _output_tensor_list[i]->set_mem_handle(output_ptr[i]);
+    
+    return &_output_tensor_list;
```

Comment:
In case of framework integration: how does it copy to framework tensor?

Reply:
We would return the tensor to user and in the iterator we call a `tensor.copy_data(user_buffer)` where we pass a user buffer and the contents of tensor is copied to user buffer. This API is added as a member function of rocalTensor class in [tensor.cpp](https://github.com/LakshmiKumar23/MIVisionX/blob/a321eeb158deb9789b36104637b5538dfe7a985b/rocAL/rocAL/source/pipeline/tensor.cpp#L318)

Reply:
Hi Fionna, In that case we need to make copy_data() uses optimized code to get the best performance